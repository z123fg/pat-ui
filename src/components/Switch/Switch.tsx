import React, { FC, useEffect, useState } from 'react';

export enum SwitchColorType {
    Primary,
    Secondary,
}

export enum SwitchSizeType {
    Medium,
    Small,
}

interface ISwitchProps {
    color?: SwitchColorType;
    size?: SwitchSizeType;
    label?: string;
    disabled?: boolean;
    checked?: boolean;
    onChange?: (e?: React.MouseEvent) => void;
}

const Switch: FC<ISwitchProps> = ({
    color = SwitchColorType.Primary,
    size = SwitchSizeType.Medium,
    label,
    disabled = false,
    checked = false,
    onChange,
}) => {
    const [checkedState, setCheckedState] = useState<boolean>(checked);

    useEffect(() => {
        setCheckedState(checked);
    }, [checked]);

    const handleClick = (e: React.MouseEvent) => {
        if (disabled) return;

        setCheckedState(!checkedState);
        onChange?.(e);
    };

    return (
        <div
            style={generateSwitchContainerStyles(size)}
        >
            <div
                className='track1'
                data-testid='switch-track'

                style={generateTrackStyles(size, disabled, color, checkedState)}
            >
                <div
                    className='ripple_container'
                    data-testid="switch-ripple"
                    style={
                        generateRippleContainerStyles(checkedState, size,color)

                    }
                >
                    <div
                        onClick={handleClick}
                        data-testid='switch-knob'

                        className='nob'
                        style={generateNobStyles(size, color, checkedState)}
                    ></div>
                </div>
            </div>
            {label !== undefined && <label>{label}</label>}
        </div>
    );
};

const generateSwitchContainerStyles = (size:SwitchSizeType) =>{
    return {
        display: 'inline-flex',
                flexDirection: 'row',
                gap: size===SwitchSizeType.Medium?'12px':'7px',
                alignItems: 'center',
    } as any
}

const generateNobStyles = (size:SwitchSizeType, color:SwitchColorType,checkedState:boolean) => {
    const onColor = color=== SwitchColorType.Primary? '#3F51B5' : '#F50057'
    return {
        width: size === SwitchSizeType.Medium ? '20px' : '16px',
        height: size === SwitchSizeType.Medium ? '20px' : '16px',
        borderRadius: '10px',
        position: 'absolute',
        background: checkedState?onColor:'#FAFAFA',
        transform: 'translate(-50%, -50%)',
        top: '50%',
        left: '50%',
    } as any
}

const generateTrackStyles = (size: SwitchSizeType, disabled: boolean, color:SwitchColorType, checkedState:boolean) => {
    const onColor = color=== SwitchColorType.Primary? 'rgba(63, 81, 181, 0.5)' : 'rgba(245, 0, 87, 0.5)'
    return {
        width: size === SwitchSizeType.Medium ? '34px' : '26px',
        height: size === SwitchSizeType.Medium ? '14px' : '10px',
        position: 'relative',
        borderRadius: '7px',
        background: disabled ? 'rgba(0,0,0,0.12)' :checkedState? onColor:'rgba(0,0,0,0.38)',
    } as any
}

const generateRippleContainerStyles = (
    checkedState: boolean,
    size: SwitchSizeType,
    color:SwitchColorType

) => {
    const leftPostion = size === SwitchSizeType.Medium ? '7px' : '5px';
    const rightPosition = size === SwitchSizeType.Medium ? '27px' : '21px';
    return {
        position: 'absolute',
        height: size === SwitchSizeType.Medium ? '38px' : '24px',
        width: size === SwitchSizeType.Medium ? '38px' : '24px',
        borderRadius: '50%',
        background: !checkedState?'rgba(0, 0, 0, 0.04)': color===SwitchColorType.Primary?'rgba(63, 81, 181, 0.08)':'rgba(245, 0, 87, 0.08)',

        transform: 'translate(-50%, -50%)',
        top: '50%',
        transition:
            'left 500ms cubic-bezier(0.4, 0, 0.2, 1) 0ms, transform 150ms cubic-bezier(0.4, 0, 0.2, 1) 0ms',
        left: !checkedState ? leftPostion : rightPosition,
    } as any;
};

export default Switch;
