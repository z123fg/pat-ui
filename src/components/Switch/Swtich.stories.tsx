import React from 'react';
import Switch, { SwitchColorType, SwitchSizeType } from './Switch';

export default {
    title: 'Switch',
    component: Switch
}

export const DefaultSwitch = () => {
    return (
        <Switch/>
    )
}


export const DiffColorSwitch = () => {
    return (
        <div>
            <h1>Primary</h1>
            <Switch/>
            <h1>Secondary</h1>
            <Switch color={SwitchColorType.Secondary}/>
        </div>
    )
}

export const DiffSizeSwitch = () => {
    return (
        <div>
            <h1>Medium</h1>
            <Switch/>
            <h1>Small</h1>
            <Switch size={SwitchSizeType.Small}/>
        </div>
    )
}

export const DisabledSwitch = () => {
    return (
        <div>
            <h1>Normal</h1>
            <Switch/>
            <h1>Disabled</h1>
            <Switch disabled/>
        </div>
    )
}

export const LabeledSwitch = () => {
    return (
        <div>
            <h1>Switch with label</h1>
            <Switch label='label'/>
        </div>
    )
}