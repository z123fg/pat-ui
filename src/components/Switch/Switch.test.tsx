import React from 'react';

import { fireEvent, render } from '@testing-library/react';
import Switch, { SwitchColorType } from './Switch';

describe('Switch component should work as expected in different conditions', ()=>{
    it('Switch component should match with the snapshop',()=>{
        const wrapper = render(<Switch/>);
        expect(wrapper).toMatchSnapshot();
    });

    it('Switch knob will change its color to Secondary if color prop is set to Secondary',()=>{
        const wrapper = render(<Switch color={SwitchColorType.Secondary}/>);
        const KnobEl = wrapper.queryByTestId('switch-knob');
        expect(KnobEl).toBeInTheDocument();
        expect(KnobEl).toHaveStyle('background: #FAFAFA');
        KnobEl&&fireEvent.click(KnobEl);
        expect(KnobEl).toHaveStyle('background: #F50057');
    });

    it('Switch should be disabled when disabled prop is true',()=>{
        const mockOnChange = jest.fn();
        const wrapper = render(<Switch disabled onChange={mockOnChange}/>);
        const TrackEl = wrapper.queryByTestId('switch-track');
        const KnobEl = wrapper.queryByTestId('switch-knob');
        expect(TrackEl).toBeInTheDocument();
        expect(TrackEl).toHaveStyle('background:rgba(0,0,0,0.12)');

        KnobEl&&fireEvent.click(KnobEl);
        expect(mockOnChange.mock.calls.length).toBe(0);
    });

    it('Switch will be turned on when checked state is true',()=>{
        const wrapper = render(<Switch/>);
        const RippleEl = wrapper.queryByTestId('switch-ripple');

        expect(RippleEl).toHaveStyle('left:7px');

        wrapper.rerender(<Switch checked/>);
        expect(RippleEl).toHaveStyle('left:27px');
    });
})