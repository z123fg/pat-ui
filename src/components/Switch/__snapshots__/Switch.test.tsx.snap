// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Switch component should work as expected in different conditions Switch component should match with the snapshop 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        style="display: inline-flex; flex-direction: row; gap: 12px; align-items: center;"
      >
        <div
          class="track1"
          data-testid="switch-track"
          style="width: 34px; height: 14px; position: relative; border-radius: 7px; background: rgba(0, 0, 0, 0.38);"
        >
          <div
            class="ripple_container"
            data-testid="switch-ripple"
            style="position: absolute; height: 38px; width: 38px; border-radius: 50%; background: rgba(0, 0, 0, 0.04); transform: translate(-50%, -50%); top: 50%; transition: left 500ms cubic-bezier(0.4, 0, 0.2, 1) 0ms, transform 150ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; left: 7px;"
          >
            <div
              class="nob"
              data-testid="switch-knob"
              style="width: 20px; height: 20px; border-radius: 10px; position: absolute; background: rgb(250, 250, 250); transform: translate(-50%, -50%); top: 50%; left: 50%;"
            />
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      style="display: inline-flex; flex-direction: row; gap: 12px; align-items: center;"
    >
      <div
        class="track1"
        data-testid="switch-track"
        style="width: 34px; height: 14px; position: relative; border-radius: 7px; background: rgba(0, 0, 0, 0.38);"
      >
        <div
          class="ripple_container"
          data-testid="switch-ripple"
          style="position: absolute; height: 38px; width: 38px; border-radius: 50%; background: rgba(0, 0, 0, 0.04); transform: translate(-50%, -50%); top: 50%; transition: left 500ms cubic-bezier(0.4, 0, 0.2, 1) 0ms, transform 150ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; left: 7px;"
        >
          <div
            class="nob"
            data-testid="switch-knob"
            style="width: 20px; height: 20px; border-radius: 10px; position: absolute; background: rgb(250, 250, 250); transform: translate(-50%, -50%); top: 50%; left: 50%;"
          />
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
